File: examples/python/03_linear_regression.py
"""
03 - Bayesian Linear Regression

Recover known regression coefficients from synthetic data using
LinearPredictor. The model is:

    y = X @ beta + epsilon,   epsilon ~ Normal(0, sigma)

Priors: beta ~ Normal(0, 5),  sigma ~ HalfNormal(2).

Note: sigma is sampled in log space; apply exp() to interpret.
"""

import numpy as np
import bayesiangpu as bg

# --- Synthetic data ---
np.random.seed(42)
N, P = 20, 3
true_beta = np.array([1.5, -0.8, 0.3])
true_sigma = 0.5

X = np.random.randn(N, P)
y = X @ true_beta + np.random.randn(N) * true_sigma

# --- Model ---
model = bg.Model()
model.param("beta", bg.Normal(0, 5), size=P)
model.param("sigma", bg.HalfNormal(2))
model.observe(bg.Normal(bg.LinearPredictor(X, "beta"), "sigma"), y.tolist())

# --- Inference ---
result = bg.sample(model, num_samples=500, num_chains=2, seed=42)

# --- Results ---
print("=" * 55)
print("Bayesian Linear Regression")
print("=" * 55)
print(f"Data: N={N}, P={P}")
print()
print(f"{'Coefficient':<14} {'True':>8} {'Estimated':>10} {'Std':>8}")
print("-" * 44)

for i in range(P):
    s = result.summarize(f"beta[{i}]")
    print(f"{'beta[' + str(i) + ']':<14} {true_beta[i]:>8.3f} {s.mean:>10.3f} {s.std:>8.3f}")

# sigma in log space
sigma_samples = np.exp(np.array(result.get_samples("sigma")))
print(f"{'sigma':<14} {true_sigma:>8.3f} {np.mean(sigma_samples):>10.3f} {np.std(sigma_samples):>8.3f}")

print()
print(f"Converged: {result.is_converged()}")

File: examples/r/03_linear_regression.R
# 03_linear_regression.R -- Linear regression with LinearPredictor
#
# Recover intercept and slope from synthetic data:
#   y = beta0 + beta1 * x + noise
#
# Priors:
#   beta  ~ Normal(0, 10)   (vector of length 2)
#   sigma ~ HalfNormal(5)
# Likelihood:
#   y ~ Normal(LinearPredictor(X, "beta"), sigma)

library(bayesiangpu)

# --- Generate synthetic data ------------------------------------------------
set.seed(321)
n <- 80
true_beta <- c(2.0, -0.75)   # intercept, slope
x <- rnorm(n)
y <- true_beta[1] + true_beta[2] * x + rnorm(n, sd = 0.5)

# Design matrix with intercept column
X <- cbind(1, x)

cat("=== Linear Regression (LinearPredictor) ===\n")
cat(sprintf("True beta = [%.2f, %.2f], noise sd = 0.50\n",
            true_beta[1], true_beta[2]))

# --- Build model ------------------------------------------------------------
model <- Model$new()
model$param("beta", Normal(0, 10), size = 2)
model$param("sigma", HalfNormal(5))
model$observe(Normal(LinearPredictor(X, "beta"), "sigma"), y)

print(model)

# --- Sample -----------------------------------------------------------------
result <- bg_sample(model, num_samples = 2000, num_chains = 4, seed = 42)
print(result)

# --- Compare ----------------------------------------------------------------
cat("\n--- True vs Estimated ---\n")
for (i in seq_along(true_beta)) {
  pname <- sprintf("beta[%d]", i - 1)
  s <- result$summarize(pname)
  cat(sprintf("beta[%d]: true = %6.2f  |  est = %6.3f  [%6.3f, %6.3f]\n",
              i - 1, true_beta[i], s$mean, s$q025, s$q975))
}
sigma_s <- result$summarize("sigma")
cat(sprintf("sigma:   true =   0.50  |  est = %6.3f  [%6.3f, %6.3f]\n",
            sigma_s$mean, sigma_s$q025, sigma_s$q975))
